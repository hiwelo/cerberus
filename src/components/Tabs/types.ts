import { ReactNode, RefObject } from 'react';

/** Defines a tab to display in a Tab list */
export type Tab = {
  /** Label of the tab */
  label: string;
  /** Content of the tab */
  content: ReactNode;
  /** React ref for the tab panel */
  refPanel?: RefObject<HTMLDivElement>;
};

/** Collection of tabs */
type TabListDictionary = {
  [key: string]: Tab;
};

export interface TabsProps {
  /** Tabs to display in the tab list */
  tabs: TabListDictionary;
}
