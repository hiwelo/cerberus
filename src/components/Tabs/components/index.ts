export { default as Container } from './Container';
export { default as TabButton } from './TabButton';
export { default as TabList } from './TabList';
export { default as TabPanel } from './TabPanel';
