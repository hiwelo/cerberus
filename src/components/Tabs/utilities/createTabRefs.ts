import { useRef } from 'react';
import { TabsProps } from '../types';

/**
 * Creates a series of React Ref for each panel part of the Tabs defined for
 * this `Tabs` component
 *
 * @param tabs The dictionary of Tabs for the related Tabs component
 */
export function createTabRefs(tabs: TabsProps['tabs']): TabsProps['tabs'] {
  // Transforms the TabListDictionary as an array of tabs
  const tabsEntries = Object.entries(tabs);
  const tabsWithRefs: TabsProps['tabs'] = {};

  // Loops over each tab and creates a React Ref for each panel HTMLElement
  tabsEntries.forEach(([tabKey, tabInfo]) => {
    const refPanel = useRef<HTMLDivElement>(null);

    tabsWithRefs[tabKey] = { ...tabInfo, refPanel };
  });

  return tabsWithRefs;
}
