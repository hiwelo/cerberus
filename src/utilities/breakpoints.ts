/** List of breakpoints in use in our styled components */
export const breakpoints = {
  xs: '0px',
  md: '540px',
};
