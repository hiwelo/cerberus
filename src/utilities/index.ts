export { breakpoints } from './breakpoints';
export { listNodesFromConnection } from './listNodesFromConnection';
