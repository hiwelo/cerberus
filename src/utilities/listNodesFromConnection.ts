/** Describes the generic structure of a GraphQL edge */
type GenericEdge<T> = {
  /** Item from a GraphQL connection object */
  node: T;
  /** Potential typing for the edge added by GraphQL */
  __typename: string;
};

/** Describes the generic structure of a GraphQL connection object */
type GenericConnection<T> = {
  edges: GenericEdge<T>[];
};

/**
 * Returns an array of all nodes contained in the provided GraphQL connection
 * object
 *
 * @param connection GraphQL connection object containing the nodes to return
 */
export function listNodesFromConnection<T>(
  connection?: GenericConnection<T>,
): T[] | undefined {
  return connection?.edges.map(({ node }) => node);
}
