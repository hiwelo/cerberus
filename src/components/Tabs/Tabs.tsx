import React, {
  FunctionComponent,
  MouseEvent,
  useEffect,
  useState,
} from 'react';
import { Container, TabButton, TabList, TabPanel } from './components';
import { TabsProps } from './types';

export const Tabs: FunctionComponent<TabsProps> = ({ tabs }) => {
  const tabList = Object.entries(tabs);
  const [firstTabKey] = tabList[0];
  const [active, setActive] = useState(firstTabKey);

  /** Event callback updating the active tab and moving the focus to it */
  const openTab = (event: MouseEvent<HTMLButtonElement>): void => {
    const sourceElement = event.currentTarget;
    const sourceId = sourceElement.id;

    // updates the active element
    setActive(sourceId);
  };

  /** Moves the focus to the tab content everytime the active element change */
  useEffect(() => {
    const activeElement = document.querySelector<HTMLButtonElement>(
      `#${active}`,
    );
    const controlledTabElement = document.querySelector<HTMLDivElement>(
      `#${activeElement?.getAttribute('aria-controls')}`,
    );

    // early-termination if the activeElement is not yet rendered
    if (!activeElement || !controlledTabElement) return;

    // early-termination if the activeElement is not the currently focused element
    if (document.activeElement !== activeElement) return;

    // moves the focus to the controlled tab
    controlledTabElement?.focus();
  }, [active]);

  return (
    <Container>
      <TabList>
        {tabList.map(([key, tabItem]) => (
          <TabButton
            key={key}
            id={key}
            aria-controls={`${key}-tab`}
            aria-expanded={active === key}
            onClick={openTab}
          >
            {tabItem.label}
          </TabButton>
        ))}
      </TabList>
      {tabList.map(([key, tabItem]) => (
        <TabPanel
          key={key}
          id={`${key}-tab`}
          hidden={active !== key ?? undefined}
          role="region"
          tabIndex={0}
          aria-labelledby={key}
        >
          {tabItem.content}
        </TabPanel>
      ))}
    </Container>
  );
};
