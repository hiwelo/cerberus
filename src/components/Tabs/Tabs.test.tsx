import React from 'react';
import { testA11y, testSnapshot } from 'utilities/test';
import { Tabs } from './Tabs';

const TestContainer = (
  <Tabs
    tabs={{
      firstTab: {
        label: 'First tab',
        content: <div>Test panel 1</div>,
      },
      secondTab: {
        label: 'Second tab',
        content: <div>Test panel 2</div>,
      },
    }}
  >
    Test
  </Tabs>
);

describe('Tabs', () => {
  testSnapshot(TestContainer);
  testA11y(TestContainer);
});
