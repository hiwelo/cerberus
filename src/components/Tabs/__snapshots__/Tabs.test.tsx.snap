// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tabs should match snapshot 1`] = `
.c1 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  padding: 1.2rem 1.6rem;
  background: none;
  border: 0;
  font-size: 1.8rem;
  font-weight: 350;
  -webkit-transition: font-weight 0.2s linear;
  transition: font-weight 0.2s linear;
  z-index: 1;
}

.c1:focus {
  outline: none;
}

.c1:hover {
  cursor: pointer;
}

.c1:focus,
.c1:hover {
  font-weight: 600;
}

.c1[aria-expanded='true'] {
  font-weight: 600;
}

.c1:first-of-type {
  position: relative;
}

.c1:first-of-type::before {
  position: absolute;
  bottom: 0;
  left: 0;
  right: 0;
  top: 0;
  display: block;
  content: '';
  background-color: #8ab0ab;
  border-radius: 0.6rem;
  box-shadow: 0 0rem 0.2rem rgba(0,0,0,0.1);
  -webkit-transition: -webkit-transform 0.5s cubic-bezier(0.175,0.885,0.32,1.045);
  -webkit-transition: transform 0.5s cubic-bezier(0.175,0.885,0.32,1.045);
  transition: transform 0.5s cubic-bezier(0.175,0.885,0.32,1.045);
  z-index: -1;
}

.c1:first-of-type[aria-expanded='false']::before {
  -webkit-transform: translateY(100%);
  -ms-transform: translateY(100%);
  transform: translateY(100%);
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  padding: 0.4rem;
  border: 0.3rem solid #8ab0ab;
  border-radius: 1.2rem;
  box-shadow: 0 0 0.1rem rgba(0,0,0,0.1) inset;
  z-index: 0;
}

.c2:focus {
  outline: none;
}

@media (min-width:540px) {
  .c1:first-of-type[aria-expanded='false']::before {
    -webkit-transform: translateX(100%);
    -ms-transform: translateX(100%);
    transform: translateX(100%);
  }
}

@media (min-width:540px) {
  .c0 {
    -webkit-flex-direction: row;
    -ms-flex-direction: row;
    flex-direction: row;
  }
}

<div
  class=""
>
  <div
    class="c0"
  >
    <button
      aria-controls="firstTab-tab"
      aria-expanded="true"
      class="c1"
      id="firstTab"
    >
      First tab
    </button>
    <button
      aria-controls="secondTab-tab"
      aria-expanded="false"
      class="c1"
      id="secondTab"
    >
      Second tab
    </button>
  </div>
  <div
    aria-labelledby="firstTab"
    class="c2"
    id="firstTab-tab"
    role="region"
    tabindex="0"
  >
    <div>
      Test panel 1
    </div>
  </div>
  <div
    aria-labelledby="secondTab"
    class="c2"
    hidden=""
    id="secondTab-tab"
    role="region"
  >
    <div>
      Test panel 2
    </div>
  </div>
</div>
`;
